---
title: "poly-A_QC"
author: "Benyu Zhou"
date: '2022-12-19'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 

```{r}
library(DESeq2)
library(tidyverse)
library(data.table)
library(dplyr)
library(stringr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(edgeR)
library(limma)
library(ggfortify)
library(mixOmics)
library(PLSDAbatch)
```



```{r}

b0<-read.table('/restricted/projectnb/casa/bzhou7/batch0/listqc_unique.txt',sep='\n')
list_sample_b0<-b0$V1
b1<-read.table('/restricted/projectnb/casa/bzhou7/batch1/listqc_unique.txt',sep='\n')
list_sample_b1<-b1$V1
b2<-read.table('/restricted/projectnb/casa/bzhou7/batch2/listqc_unique.txt',sep='\n')
list_sample_b2<-b2$V1
b3<-read.table('/restricted/projectnb/casa/bzhou7/batch3/listqc_unique.txt',sep='\n')
list_sample_b3<-b3$V1
b4<-read.table('/restricted/projectnb/casa/bzhou7/batch4/listqc_unique.txt',sep='\n')
list_sample_b4<-b4$V1
b5<-read.table('/restricted/projectnb/casa/bzhou7/batch5/listqc_unique.txt',sep='\n')
list_sample_b5<-b5$V1

b0$batch<-0
b1$batch<-1
b2$batch<-2
b3$batch<-3
b4$batch<-4
b5$batch<-5


total<-rbind(b1,b2,b3,b4,b5,b0)



all_plus<- total%>%filter(grepl("plus", V1))
all_plus<-all_plus%>%separate(V1, c('Sample', 'Identifier'),sep='-')
all_minus<-total%>%filter(grepl("minus", V1))
all_minus<-all_minus%>%separate(V1, c('Sample', 'Identifier'),sep='-')
deserted_samples<-c('1071-minus_182_011_S130', '1071-plus_007_186_S158', '1073-plus_364_213_S121',
                    '1080-plus_292_285_S105', '1113-plus_325_252_S125')
total_updated<-total[-which(total$V1%in%deserted_samples),]
plus_updated<- total_updated%>%filter(grepl("plus", V1))
plus_updated<-plus_updated%>%separate(V1, c('Sample', 'Identifier'),sep='-')
minus_updated<-total_updated%>%filter(grepl("minus", V1))
minus_updated<-minus_updated%>%separate(V1, c('Sample', 'Identifier'),sep='-')

plus_samples<-unique(all_plus$Sample)
minus_samples<-unique(all_minus$Sample)
missing_plus<-minus_samples[!(minus_samples %in% plus_samples)]
missing_minus<-plus_samples[!(plus_samples %in% minus_samples)]

ind<-which(table(all_plus$Sample)>1)
more_than_once_plus<-table(all_plus$Sample)[ind]
ind<-which(table(all_minus$Sample)>1)
more_than_once_minus<-table(all_minus$Sample)[ind]
more_than_once_plus_df<-all_plus[all_plus$Sample %in% names(more_than_once_plus),]%>%arrange(batch)
more_than_once_minus_df<-all_minus[all_minus$Sample %in% names(more_than_once_minus),]%>%arrange(batch)

list_sample_b0<-list_sample_b0[-which(list_sample_b0%in%deserted_samples)]
list_sample_b2<-list_sample_b2[-which(list_sample_b2%in%deserted_samples)]
list_sample_b3<-list_sample_b3[-which(list_sample_b3%in%deserted_samples)]

```


```{r}
#Using functions
func1 <- function(x,c){
  result <- unlist(strsplit(x[c],split="_"))[1]
  return(result)
}

func2 <- function(x,c){
  result <- unlist(strsplit(x[c],split="_"))[2]
  return(result)
}


all_batches<-rbind(plus_updated,minus_updated)%>%arrange(Sample)%>%
  separate(Identifier, c('Strand', 'Identifier'),sep='(?<=[a-z])\\_(?=[0-9])')
```


```{r}
#Build sample list
list_df_0 <- rep(NA, 0)
list_df_1 <- rep(NA, 0)
list_df_2 <- rep(NA, 0)
list_df_3 <- rep(NA, 0)
list_df_4 <- rep(NA, 0)
list_df_5 <- rep(NA, 0)


for (i in list_sample_b0){
  list_df_0 <- append(list_df_0,paste('/restricted/projectnb/casa/bzhou7/batch0/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b1){
  list_df_1 <- append(list_df_1,paste('/restricted/projectnb/casa/bzhou7/batch1/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b2){
  list_df_2 <- append(list_df_2,paste('/restricted/projectnb/casa/bzhou7/batch2/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b3){
  list_df_3 <- append(list_df_3,paste('/restricted/projectnb/casa/bzhou7/batch3/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b4){
  list_df_4 <- append(list_df_4,paste('/restricted/projectnb/casa/bzhou7/batch4/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b5){
  list_df_5 <- append(list_df_5,paste('/restricted/projectnb/casa/bzhou7/batch5/lrna/',i,'_test.genes.results', sep = ''))
}
# 
# 
# 
# #ribosome data
# ribo_data<- read.csv("/restricted/projectnb/casa/bu_brain_rnaseq/hippo_rawdata/Metadata/BuBrain_RNAseq.pheno.210s.small.new.v3.csv",header = TRUE, sep = '\t')
# ribo_data<-ribo_data%>%filter(RNAseqId%in%all_batches$Sample)
# 


```


```{r}
total_df<-data.frame()


for (i in list_df_1[1]) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=1
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(df,total_df)
}

for (i in list_df_1[2:46]) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=1
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_2) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=2
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_3) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=3
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_4) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=4
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_5) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=5
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

total_df_wo_b0<-total_df

for (i in list_df_0) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=0
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

df_pca_wb0<-total_df[,which(apply(total_df, 2, var)!=0)]
df_pca_wb0$strand<-total_df$strand
df_pca_wb0$strand<-as.factor(df_pca_wb0$strand)
df_pca_wb0$batch<-as.factor(df_pca_wb0$batch)
```


```{r}
library(ggfortify)
library(mixOmics)
library(PLSDAbatch)
```


```{r}
df_pca_plus<- total_df[total_df$strand=="plus",]
df_pca_plus<-df_pca_plus[,which(apply(df_pca_plus, 2, var)!=0)]
df_pca_minus<- total_df[total_df$strand=="minus",]
df_pca_minus<-df_pca_minus[,which(apply(df_pca_minus, 2, var)!=0)]
```
```{r}
rownames(df_pca_minus)<-minus_updated$Sample
rownames(df_pca_plus)<-plus_updated$Sample
```


```{r}
# df_pca_minus[,1:47822]<-log(df_pca_minus[,1:47822]+0.1)
# df_pca_plus[,1:44778]<-log(df_pca_plus[,1:44778]+0.1)
# plus.pca<-pca(df_pca_plus[,1:42339], center = TRUE, scale =TRUE)
# minus.pca<-pca(df_pca_minus[,1:44883], center = TRUE, scale =TRUE)


write.table(df_pca_minus,"df_pca_minus.csv",row.names = TRUE,sep=",")
write.table(df_pca_plus,"df_pca_plus.csv",row.names = TRUE,sep=",")

#SAVED
# ##
# df_pca_minus<-read.csv("df_pca_minus.csv")
# df_pca_plus<-read.csv("df_pca_plus.csv")
```


```{r}
df_pca_minus<-read.csv("df_pca_minus.csv",header = TRUE)
df_pca_plus<-read.csv("df_pca_plus.csv",header = TRUE)

log_offset <- 0.0001
df_pca_minus_log<-log10(df_pca_minus[,1:47822]+log_offset)
df_pca_plus_log<-log10(df_pca_plus[,1:44778]+log_offset)

plus.pca<-pca(df_pca_plus_log, center = TRUE, scale =FALSE)
# plus.pca.scaled<-pca(df_pca_plus_log[,1:42339], center = TRUE, scale =TRUE)
minus.pca<-pca(df_pca_minus_log, center = TRUE, scale =FALSE)
# minus.pca.scaled<-pca(df_pca_minus_log[,1:44883], center = TRUE, scale =TRUE)

plus.pca.plot<- 
  Scatter_Density(object = plus.pca, 
                  batch = df_pca_plus$batch, 
                  title = 'Poly-A plus strand data PCA plot', 
                  batch.legend.title = 'batch')
minus.pca.plot<- 
  Scatter_Density(object = minus.pca, 
                  batch = df_pca_minus$batch, 
                  title = 'Poly-A minus strand data PCA plot', 
                  batch.legend.title = 'batch')

```
#Traditional standard deviaition outliers identification based on visualization


```{r}
first_2_minus<-minus.pca$variates$X
first_2_plus<-plus.pca$variates$X
```

```{r}
first_2_minus<-as.data.frame(first_2_minus)
ind<-which(first_2_minus$PC1< (-50) & first_2_minus$PC2>30)
minus.small.cluster<-first_2_minus[ind,]
minus.small.cluster.df<-minus_updated[ind,]
```

```{r}
minus.pca.outliers<-apply(first_2_minus, 2, function(x) which( abs(x - mean(x)) > (2 * sd(x)) ))
plus.pca.outliers<-apply(first_2_plus, 2, function(x) which( abs(x - mean(x)) > (2 * sd(x)) ))
#Record outliers in each PC, and find outliers that appear in both PC
minus.pc1.outliers<-minus.pca.outliers$PC1
minus.pc2.outliers<-minus.pca.outliers$PC2
plus.pc1.outliers<-plus.pca.outliers$PC1
plus.pc2.outliers<-plus.pca.outliers$PC2
ind<-which(names(plus.pc2.outliers)%in%names(plus.pc1.outliers))
plus.outliers<-names(plus.pc2.outliers)[ind]
ind<-which(names(minus.pc2.outliers)%in%names(minus.pc1.outliers))
minus.outliers<-names(minus.pc2.outliers)[ind]
minus_outliers_df<-minus_updated[minus.outliers,]
plus_outliers_df<-plus_updated[plus.outliers,]
```


```{r}
#Updated PCA
df_pca_minus_updated<-df_pca_minus[-as.numeric(minus.outliers),]
df_pca_plus_updated<-df_pca_plus[-as.numeric(plus.outliers),]
df_pca_minus_updated_log<-df_pca_minus_log[-as.numeric(minus.outliers),]
df_pca_plus_updated_log<-df_pca_plus_log[-as.numeric(plus.outliers),]
```


```{r}
plus.pca_updated<-pca(df_pca_plus_updated_log, center = TRUE, scale =FALSE)

minus.pca_updated<-pca(df_pca_minus_updated_log, center = TRUE, scale =FALSE)

plus.pca.plot_updated<- 
  Scatter_Density(object = plus.pca_updated, 
                  batch = df_pca_plus_updated$batch, 
                  title = 'Updated Poly-A plus strand data PCA plot', 
                  batch.legend.title = 'batch')
minus.pca.plot_updated<- 
  Scatter_Density(object = minus.pca_updated, 
                  batch = df_pca_minus_updated$batch, 
                  title = 'Updated Poly-A minus strand data PCA plot', 
                  batch.legend.title = 'batch')




# plus_multiqc[which(plus_multiqc$plus_star.uniquely_mapped_percent<50),]
# minus_multiqc[which(minus_multiqc$minus_star.uniquely_mapped_percent<50),]
```
#Outliers identification option 2: 
Outlier detection by the Mahalanobis distances of PC1 and PC2. Also plot PC1 and PC2 with its confidence ellipse. (package mt)

```{r}
#install.packages('mt')
library('mt')
```

```{r}
x<-pca.outlier(df_pca_minus_log, center = TRUE, scale =FALSE, conf.level = 0.975) 
x$outlier
x$plot
#Comment: Looks nice, actually got 2 more outliers than the 2 sd away and finding overlaps. Could be a handy way to identify outliers
```
```{r}
p<-pca.outlier(df_pca_plus_log, center = TRUE, scale =FALSE, conf.level = 0.975) 
p$outlier
```
#Option 3 LOF (LOCAL OUTLIER FACTOR)
K nearest neighboghs based, seem to not really good for pca outliers detection, in this case when clearly cluster separation appeared on the plot. But can be good to find outliers in each cluster.

```{r}
library('bigutilsr')
```

```{r}
U3<-first_2_plus
llof <- LOF(U3)  # log(LOF) by default
#qplot(dist2, llof)
cowplot::plot_grid(
  qplot(U3[, 1], U3[, 2], color = llof, size = I(3)) + coord_equal() + 
    scale_color_viridis_c(breaks = NULL),
  rel_widths = c(0.7, 0.4), scale = 0.95
)
```

#option4:checking 3 statistics RLE，Hierarchical clustering，D-statistic(XQTL PIPELINE)
```{r}
    # Load packages
    library(dplyr)
    library(readr)
    library(tibble)
    library(purrr)
    ## Setting inputs
    low_expr_TPM <- 0.1
    low_expr_TPM_percent <- 0.2
    ## QC
    ### Data integrity check. The input file should has sample ID as column name; the first column is gene ID
    TPM_data<-transpose(df_pca_minus[,1:47822])
    rownames(TPM_data)<-colnames(df_pca_minus_log)
    colnames(TPM_data)<-rownames(df_pca_minus)
    ##################################################
    #Check if duplication, but seems over complicated    
    TPM_data$feature<-rownames(TPM_data)
    TPM_data<-TPM_data %>%relocate(feature)
    if( sum(duplicated(TPM_data$feature)) > 0){
      message("feature (e.g. gene names) should be in the first column. Please remove duplicated feature IDs, Exit!")
      quit(save = "no", status = 1, runLast = FALSE)
    }else{
      rownames(TPM_data) <- TPM_data$feature
      TPM_data = TPM_data[,-1]
      loaded_sample_count <- ncol(TPM_data)
    }
    ####################################################
    #### NA check
    if(sum(is.na(TPM_data)) > 0 ){      
      message(paste0("NA is not allowed in the data, there are ",sum(is.na(TPM_data))," NAs, Exit!"))
      quit(save = "no", status = 1, runLast = FALSE)}
    # #### make sure every expr column is in numeric type
    # matrix_check <- map(TPM_data, is.numeric) %>% unlist
    # if(sum(!matrix_check) > 0){
    #   message("The following column(s) in expression matrix is/are NOT in numeric type. Plese check, Proceed by excluding those samples")
    #   message(paste(names(matrix_check)[!matrix_check], collapse = "; "))
    #   TPM_data = TPM_data[,matrix_check]
    # }
    # message("Gene expression profiles loaded successfully!")
    # message(paste(nrow(TPM_data), "genes and", ncol(TPM_data), "samples are loaded from", $[_input:r], sep = " "))
    
    #### Filter out low exp genes
    keep_genes_idx <- (rowMeans(TPM_data>low_expr_TPM)>low_expr_TPM_percent) 
    TPM_data = TPM_data[keep_genes_idx,]
    message(paste(sum(1 - keep_genes_idx), "genes are filtered by filter: >", low_expr_TPM_percent*100, "% samples have expression values <", low_expr_TPM))
    message(paste(sum(keep_genes_idx), "genes left, saving output to $[_output]."))
    TPM_data%>%as_tibble(rownames = "gene_ID")%>%write_delim("$[_output]","\t")
```

```{r}
library(RColorBrewer)
    library(ape)
    library(reshape2)
    library(dplyr)
    library(readr)
    ## Setting parameters
    RLEFilterPercent <-0.05
    DSFilterPercent <- 0.05
    ### Hcluster parameter
    pvalues.cut <- 0.05
    topk_genes <- 100
    cluster_percent <- 0.6
    treesNum <- 5
    ## Define functions to accomodate rank deficient covariance matrices in https://github.com/cumc/xqtl-pipeline/issues/307
    mahalanobis = function (x, center, cov, inverted = FALSE, ...) 
        {
            x <- if (is.vector(x)) 
                matrix(x, ncol = length(x))
            else as.matrix(x)
            if (!isFALSE(center)) 
                x <- sweep(x, 2L, center)
            if (!inverted) 
                cov <- MASS::ginv(cov, ...)
            setNames(rowSums(x %*% cov * x), rownames(x))
        }
          
    ## laod Data
    # TPM_data <- read_tsv('/restricted/projectnb/casa/skandoi/bulk_qc/fastq.rnaseqc.low_expression_filtered.tpm.gct.gz', col_names = TRUE, comment = "#")%>%as.data.frame()
    # rownames(TPM_data) <- TPM_data$gene_ID
    # TPM_data = TPM_data[,-1]
    RLEFilterLength <- RLEFilterPercent*ncol(TPM_data)
    DSFilter <- DSFilterPercent*ncol(TPM_data)
```


```{r}
## Outlier detection
  
    ### RLE
    # https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L71
    log_offset <- 0.0001
    logtpm = log10(TPM_data%>%as.matrix + log_offset)
    rle=logtpm-apply(logtpm, 1, median) # change "/" to "-" so that we got log(fold-change) which centered on 0 on the RLE plot.
    iqr = apply(rle,2,IQR)
    rle=melt( rle, variable.name = "Sample",value.name ="TPM", id="ID")
    names(rle) <- c("feature","Sample","TPM")
    rle_IQR <- rle %>% group_by(Sample) %>% summarise(IQR = IQR(TPM))
    rle_IQR_range <- rle_IQR$IQR %>% range %>% abs() %>% max()
    rle_IQR_range <- 2*rle_IQR_range %>% ceiling()
    bymedian <- with(rle, reorder(Sample, TPM, IQR))  # sort by IQR
    par(mar=c(3,3,3,3))
    pdf(file = "$[_output:n].RLEplot.pdf")
    boxplot(TPM ~ bymedian, data=rle, outline=F, ylim = c(-rle_IQR_range, rle_IQR_range), las=2, boxwex=1, col='gray', cex.axis=0.3, main="RLE plot before QC", xlab="", ylab="Residual expression levels", frame=F)
    dev.off()
    ExpPerSample <- nrow(TPM_data)
    RLEFilterList <- unique(bymedian[((length(bymedian)-ExpPerSample*RLEFilterLength)+1):length(bymedian)]) #filtered
    RLEFilterList <- as.character(RLEFilterList)
    message(paste0("The right most ", RLEFilterPercent*100, "% samples (N = ", length(RLEFilterList), ") are marked as candidate outliers in this step:") )
    message(RLEFilterList)
```


```{r}
### hcluster  
    sampleDists <- 1 - cor(logtpm, method='spearman')
    hc <- hclust(as.dist(sampleDists), method = "complete")
    hcphy <- as.phylo(hc)
  
    pdf(file = "$[_output:n].preQC_cluster.pdf")
    plot(hcphy, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Sample clustering before QC (Spearman - Cor.)")
    dev.off()
    # https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L102
    log_offset <- 1
    logtpm = log10(TPM_data%>%as.matrix + log_offset)
    ntop <- topk_genes
    Pvars <- apply(logtpm, 1, var)
    select <- order(Pvars, decreasing =TRUE)[seq_len(min(ntop, length(Pvars)))]
    MD_matrix <- logtpm[select, ]
    MahalanobisDistance = mahalanobis(t(MD_matrix), colMeans(t(MD_matrix)), cov(t(MD_matrix))) 
    # Note: t(MD_matrix)=sample_row*gene_column, Manhalanobis() returns one vector with length=row number
    pvalues = pchisq(MahalanobisDistance, df=nrow(MD_matrix), lower.tail=F)
    pvalues.adjust = p.adjust(pvalues, method ="bonferroni") # adjusted pvalues for each sample
    pvalues.low <- pvalues.adjust[pvalues.adjust<pvalues.cut]
    
    HCoutliers <- character()
    for(x in c(1:treesNum)){
      trees <- cutree(hc,k=x)
      idx <- c(1:x)#which tree is checking
      for(i in idx)
      {
        group <- hc$labels[which(trees == i)]
        if(sum(group %in% names(pvalues.low))/length(group) >= cluster_percent)
        {
          HCoutliers <- union(HCoutliers,group)
        }
      }
    }
    
    message(paste(length(HCoutliers), "samples are marked as candidate outlier(s) in this step.", sep = " "))
    if(length(HCoutliers)>0){
      message("Sample outliers are marked in red as follows:")
      message(HCoutliers)
      co1 = hc$labels%in%HCoutliers
      co1[which(co1 == "FALSE")]<-"gray0"
      co1[which(co1 == "TRUE")]<-"red"
      par(mar=c(3,3,3,3))
  
    pdf(file = "$[_output:n].cluster.pdf")
    plot(hcphy, tip.col = co1, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Label Outliers in Red")
    Xcol = c("gray0", "red")
    Xtext = c("Normal Sample", "Outliers")
    legend('bottomleft',pch=21,Xtext, col='white',pt.bg=Xcol, cex=1)
    dev.off()
    }else{
      message("No outlier detected.")
    }
```


```{r}
### D-s
    D = apply(1-sampleDists, 1, median)
    pdf(file = "$[_output:n].D_stat_hist.pdf")
    hist(D, breaks=100, ylab="Number of samples", xlab="D-statistic", main="Histogram of Sample D-statistics before data QC")
    dev.off()
    
    DSFilter <- sort(D)[DSFilter]
    D<-as.data.frame(D)
    D<-data.frame(Sample = rownames(D),D = D$D)
    D_filterList = D%>%filter(D <= DSFilter)
    D_filterList <- D_filterList$Sample
    D_filterList<-as.character(D_filterList)
    message(paste0("The right most ", DSFilterPercent*100, "% samples (N=", length(D_filterList), ") are marked as candidate outliers in this step:") )
    message(D_filterList)
    
    ## Outliers are the intersect of three candidates list
    outliersList <- intersect(RLEFilterList,intersect(HCoutliers,D_filterList))
    message("Outliers:")
    message(outliersList)
    outliersIndex <- which(colnames(logtpm) %in% outliersList)
    if(!length(outliersIndex) == 0){
        TPM_data <- TPM_data[,-outliersIndex]
    }
    # ## Add 2 header lines, https://github.com/getzlab/rnaseqc/blob/286f99dfd4164d33014241dd4f3149da0cddf5bf/src/RNASeQC.cpp#L426
    # cat(paste("#1.2\n#", nrow(TPM_data), ncol(TPM_data) - 2, "\n"), file=$[_output:nr], append=FALSE)
    # TPM_data%>%as_tibble(rownames = "gene_ID")%>%write_delim($[_output:nr],delim = "\t",col_names = T, append = T)
```



#Correlation test
```{r}
####Correlation test
df_minus_t<-transpose(df_pca_minus_log)
df_minus_t<-df_minus_t[-47823,]
df_plus_t<-transpose(df_pca_plus_log)
df_plus_t<-df_plus_t[-44779,]
plus_cor<-cor(df_plus_t)
minus_cor<-cor(df_minus_t)

minus.mean_corr<-apply(minus_cor, 2, function(x) (sum(x)-1)/(length(x)-1))
minus.mean_corr<-data.frame(minus.mean_corr)
minus_mean_mean<-mean(minus.mean_corr[,1])
minus_sd_mean<-sd(minus.mean_corr[,1])
ind<-which( abs(minus.mean_corr - minus_mean_mean) > (3 * minus_sd_mean) )
minus_corr_outliers<-minus_updated[ind,]

plus.mean_corr<-apply(plus_cor, 2, function(x) (sum(x)-1)/(length(x)-1))
plus.mean_corr<-data.frame(plus.mean_corr)
plus_mean_mean<-mean(plus.mean_corr[,1])
plus_sd_mean<-sd(plus.mean_corr[,1])
ind<-which( abs(plus.mean_corr - plus_mean_mean) > (3 * plus_sd_mean))
plus_corr_outliers_df<-plus_updated[ind,]
```


```{r}
library("rtracklayer")
gtf_data = import('/restricted/projectnb/casa/jmh/ref/GRCh38/annotation/GRCh38.95/Homo_sapiens.GRCh38.95.gtf') 
gtf_data = as.data.frame(gtf_data)
gtf <- gtf_data[which(gtf_data$type=='gene'),]
colnames(gtf)[1] <- 'geneid'

all_genes<-gtf[,c('gene_name','gene_biotype')]
minus_genes<-data.frame(colnames(df_pca_minus_log))
colnames(minus_genes)<-"gene_name"
plus_genes<-data.frame((colnames(df_pca_plus_log)))
colnames(plus_genes)<-"gene_name"
```

```{r}
minus_genes<-minus_genes[-47823,]
plus_genes<-plus_genes[-44779,]
df_minus_t<-cbind(df_minus_t, minus_genes)
df_plus_t<-cbind(df_plus_t,plus_genes)
```

```{r}
df_minus_protein<-merge(df_minus_t, all_genes, by.x="minus_genes",by.y="gene_name")
df_minus_protein<-df_minus_protein[which(df_minus_protein$gene_biotype=="protein_coding"),]
#df_minus_protein<-df_minus_protein[,-c(1,152)]

df_plus_protein<-merge(df_plus_t, all_genes, by.x="plus_genes",by.y="gene_name")
df_plus_protein<-df_plus_protein[which(df_plus_protein$gene_biotype=="protein_coding"),]
#df_plus_protein<-df_plus_protein[,-c(1,145)]
```


```{r}
plus_cor_protein<-cor(df_plus_protein[,2:144])
minus_cor_protein<-cor(df_minus_protein[,2:151])

minus.mean_corr_protein<-apply(minus_cor_protein, 2, function(x) (sum(x)-1)/(length(x)-1))
minus.mean_corr_protein<-data.frame(minus.mean_corr_protein)
minus_mean_mean_protein<-mean(minus.mean_corr_protein[,1])
minus_sd_mean_protein<-sd(minus.mean_corr_protein[,1])
ind<-which(abs(minus.mean_corr_protein - minus_mean_mean_protein) > (3 * minus_sd_mean_protein) )
minus_corr_outliers_protein<-minus_updated[ind,]

plus.mean_corr_protein<-apply(plus_cor_protein, 2, function(x) (sum(x)-1)/(length(x)-1))
plus.mean_corr_protein<-data.frame(plus.mean_corr_protein)
plus_mean_mean_protein<-mean(plus.mean_corr_protein[,1])
plus_sd_mean_protein<-sd(plus.mean_corr_protein[,1])
ind<-which( abs(plus.mean_corr_protein - plus_mean_mean_protein) > (3 * plus_sd_mean_protein))
plus_corr_outliers_protein<-plus_updated[ind,]
```


```{r}
#install.packages("git2r")
library(git2r)

# Insure you have navigated to a directory with a git repo.
dir <- "/restricted/projectnb/casa/bzhou7/bulk-rna-seq-pipeline"
setwd(dir)

# Configure git.
git2r::config(user.name = "Benjo23",user.email = "bzhou7@bu.edu")

# Check git status.
gitstatus()


# Add and commit changes. 
gitadd()
gitcommit()

# Push changes to github.
gitpush()
```





