---
title: "poly-A_QC"
author: "Benyu Zhou"
date: '2022-12-19'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## packages

```{r}
library(DESeq2)
library(tidyverse)
library(data.table)
library(dplyr)
library(stringr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(edgeR)
library(limma)
library(ggfortify)
library(mixOmics)
library(PLSDAbatch)
```

#Read the sample list FOR THE POLY-A DATA

```{r}
#Read the sample list
b0<-read.table('/restricted/projectnb/casa/bzhou7/batch0/listqc_unique.txt',sep='\n')
list_sample_b0<-b0$V1
b1<-read.table('/restricted/projectnb/casa/bzhou7/batch1/listqc_unique.txt',sep='\n')
list_sample_b1<-b1$V1
b2<-read.table('/restricted/projectnb/casa/bzhou7/batch2/listqc_unique.txt',sep='\n')
list_sample_b2<-b2$V1
b3<-read.table('/restricted/projectnb/casa/bzhou7/batch3/listqc_unique.txt',sep='\n')
list_sample_b3<-b3$V1
b4<-read.table('/restricted/projectnb/casa/bzhou7/batch4/listqc_unique.txt',sep='\n')
list_sample_b4<-b4$V1
b5<-read.table('/restricted/projectnb/casa/bzhou7/batch5/listqc_unique.txt',sep='\n')
list_sample_b5<-b5$V1

b0$batch<-0
b1$batch<-1
b2$batch<-2
b3$batch<-3
b4$batch<-4
b5$batch<-5

#bind them together into a total df include all the samples
total<-rbind(b1,b2,b3,b4,b5,b0)

#separate in to minus and plus
all_plus<- total%>%filter(grepl("plus", V1))
all_plus<-all_plus%>%separate(V1, c('Sample', 'Identifier'),sep='-')
all_minus<-total%>%filter(grepl("minus", V1))
all_minus<-all_minus%>%separate(V1, c('Sample', 'Identifier'),sep='-')

#To find those samples that are missing in plus or minus
plus_samples<-unique(all_plus$Sample)
minus_samples<-unique(all_minus$Sample)
missing_plus<-minus_samples[!(minus_samples %in% plus_samples)]
missing_minus<-plus_samples[!(plus_samples %in% minus_samples)]


#Get rid of duplicated results identified as as not good quality in previous fastqc report

deserted_samples<-c('1071-minus_182_011_S130', '1071-plus_007_186_S158', '1073-plus_364_213_S121',
                    '1080-plus_292_285_S105', '1113-plus_325_252_S125')
total_updated<-total[-which(total$V1%in%deserted_samples),]
plus_updated<- total_updated%>%filter(grepl("plus", V1))
plus_updated<-plus_updated%>%separate(V1, c('Sample', 'Identifier'),sep='-')
minus_updated<-total_updated%>%filter(grepl("minus", V1))
minus_updated<-minus_updated%>%separate(V1, c('Sample', 'Identifier'),sep='-')


list_sample_b0<-list_sample_b0[-which(list_sample_b0%in%deserted_samples)]
list_sample_b2<-list_sample_b2[-which(list_sample_b2%in%deserted_samples)]
list_sample_b3<-list_sample_b3[-which(list_sample_b3%in%deserted_samples)]

```

#Using functions to split strings for gene ids
```{r}
func1 <- function(x,c){
  result <- unlist(strsplit(x[c],split="_"))[1]
  return(result)
}

func2 <- function(x,c){
  result <- unlist(strsplit(x[c],split="_"))[2]
  return(result)
}


all_batches<-rbind(plus_updated,minus_updated)%>%arrange(Sample)%>%
  separate(Identifier, c('Strand', 'Identifier'),sep='(?<=[a-z])\\_(?=[0-9])')
```

#Read all the test,gene,results into a large data frame containing all samples (total_df)
```{r}
#Build sample list
list_df_0 <- rep(NA, 0)
list_df_1 <- rep(NA, 0)
list_df_2 <- rep(NA, 0)
list_df_3 <- rep(NA, 0)
list_df_4 <- rep(NA, 0)
list_df_5 <- rep(NA, 0)


for (i in list_sample_b0){
  list_df_0 <- append(list_df_0,paste('/restricted/projectnb/casa/bzhou7/batch0/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b1){
  list_df_1 <- append(list_df_1,paste('/restricted/projectnb/casa/bzhou7/batch1/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b2){
  list_df_2 <- append(list_df_2,paste('/restricted/projectnb/casa/bzhou7/batch2/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b3){
  list_df_3 <- append(list_df_3,paste('/restricted/projectnb/casa/bzhou7/batch3/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b4){
  list_df_4 <- append(list_df_4,paste('/restricted/projectnb/casa/bzhou7/batch4/lrna/',i,'_test.genes.results', sep = ''))
}

for (i in list_sample_b5){
  list_df_5 <- append(list_df_5,paste('/restricted/projectnb/casa/bzhou7/batch5/lrna/',i,'_test.genes.results', sep = ''))
}

```


```{r}
total_df<-data.frame()


for (i in list_df_1[1]) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=1
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(df,total_df)
}

for (i in list_df_1[2:46]) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=1
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_2) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=2
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_3) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=3
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_4) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=4
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

for (i in list_df_5) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=5
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

total_df_wo_b0<-total_df

for (i in list_df_0) {
  df <- read.delim2(i)
  df$geneid <- apply(df, 1, func1, c='gene_id')
  df$genesym <- apply(df, 1, func2, c='gene_id')
  df$TPM <- as.double(df$TPM)
  df$expected_count <- as.integer(df$expected_count)
  df<-transpose(df[,c(6,9)],make.names = "genesym")
  df$batch=0
  if (grepl("plus", i, fixed = TRUE)){
    df$strand="plus"
  }
  else{
    df$strand="minus"
  }
  total_df<-rbind(total_df,df)
}

df_pca_wb0<-total_df[,which(apply(total_df, 2, var)!=0)]
df_pca_wb0$strand<-total_df$strand
df_pca_wb0$strand<-as.factor(df_pca_wb0$strand)
df_pca_wb0$batch<-as.factor(df_pca_wb0$batch)
```


```{r}
#separate them into plus and minus
df_pca_plus<- total_df[total_df$strand=="plus",]
df_pca_plus<-df_pca_plus[,which(apply(df_pca_plus, 2, var)!=0)]
df_pca_minus<- total_df[total_df$strand=="minus",]
df_pca_minus<-df_pca_minus[,which(apply(df_pca_minus, 2, var)!=0)]
```

```{r}
rownames(df_pca_minus)<-minus_updated$Sample
rownames(df_pca_plus)<-plus_updated$Sample
```


```{r}
#Save them into csv files

write.table(df_pca_minus,"df_pca_minus.csv",row.names = TRUE,sep=",")
write.table(df_pca_plus,"df_pca_plus.csv",row.names = TRUE,sep=",")

#SAVED
# ##
# df_pca_minus<-read.csv("df_pca_minus.csv")
# df_pca_plus<-read.csv("df_pca_plus.csv")
```


#In the future can directly read the csv files for tpm data
```{r}
df_pca_minus<-read.csv("df_pca_minus.csv",header = TRUE)
df_pca_plus<-read.csv("df_pca_plus.csv",header = TRUE)
```

#Minus QC
```{r}
    # Load packages
    library(dplyr)
    library(readr)
    library(tibble)
    library(purrr)
    ## Setting inputs
    low_expr_TPM <- 0.1
    low_expr_TPM_percent <- 0.2
    ## QC
    ### Data integrity check. The input file should has sample ID as column name; the first column is gene ID
    TPM_data<-transpose(df_pca_minus[,1:47822])
    rownames(TPM_data)<-colnames(df_pca_minus[,1:47822])
    colnames(TPM_data)<-rownames(df_pca_minus)
    ##################################################
    #Check if duplication, but seems over complicated    
    TPM_data$feature<-rownames(TPM_data)
    TPM_data<-TPM_data %>%relocate(feature)
    if( sum(duplicated(TPM_data$feature)) > 0){
      message("feature (e.g. gene names) should be in the first column. Please remove duplicated feature IDs, Exit!")
      quit(save = "no", status = 1, runLast = FALSE)
    }else{
      rownames(TPM_data) <- TPM_data$feature
      TPM_data = TPM_data[,-1]
      loaded_sample_count <- ncol(TPM_data)
    }
    ####################################################
    #### NA check
    if(sum(is.na(TPM_data)) > 0 ){      
      message(paste0("NA is not allowed in the data, there are ",sum(is.na(TPM_data))," NAs, Exit!"))
      quit(save = "no", status = 1, runLast = FALSE)}
    # #### make sure every expr column is in numeric type
    # matrix_check <- map(TPM_data, is.numeric) %>% unlist
    # if(sum(!matrix_check) > 0){
    #   message("The following column(s) in expression matrix is/are NOT in numeric type. Plese check, Proceed by excluding those samples")
    #   message(paste(names(matrix_check)[!matrix_check], collapse = "; "))
    #   TPM_data = TPM_data[,matrix_check]
    # }
    # message("Gene expression profiles loaded successfully!")
    # message(paste(nrow(TPM_data), "genes and", ncol(TPM_data), "samples are loaded from", $[_input:r], sep = " "))
    
    #### Filter out low exp genes
    keep_genes_idx <- (rowMeans(TPM_data>low_expr_TPM)>low_expr_TPM_percent) 
    TPM_data = TPM_data[keep_genes_idx,]
```

```{r}
log_offset <- 0.0001
TPM_t<-as.data.frame(t(TPM_data))
df_pca_minus_log<-log10(TPM_t+log_offset)
# df_pca_plus_log<-log10(TPM_t+log_offset)

# plus.pca<-pca(df_pca_plus_log, center = TRUE, scale =FALSE)
# # plus.pca.scaled<-pca(df_pca_plus_log[,1:42339], center = TRUE, scale =TRUE)
minus.pca<-pca(df_pca_minus_log, center = TRUE, scale =FALSE)
# minus.pca.scaled<-pca(df_pca_minus_log[,1:44883], center = TRUE, scale =TRUE)

# plus.pca.plot<- 
#   Scatter_Density(object = plus.pca, 
#                   batch = df_pca_plus$batch, 
#                   title = 'Poly-A plus strand data PCA plot', 
#                   batch.legend.title = 'batch')
minus.pca.plot<- 
  Scatter_Density(object = minus.pca, 
                  batch = df_pca_minus$batch, 
                  title = 'Poly-A minus strand data PCA plot', 
                  batch.legend.title = 'batch')

```

#Plus QC
```{r}
    # Load packages
    library(dplyr)
    library(readr)
    library(tibble)
    library(purrr)
    ## Setting inputs
    low_expr_TPM <- 0.1
    low_expr_TPM_percent <- 0.2
    ## QC
    ### Data integrity check. The input file should has sample ID as column name; the first column is gene ID
    TPM_data<-transpose(df_pca_plus[,1:44778])
    rownames(TPM_data)<-colnames(df_pca_plus[,1:44778])
    colnames(TPM_data)<-rownames(df_pca_plus)
    ##################################################
    #Check if duplication, but seems over complicated    
    TPM_data$feature<-rownames(TPM_data)
    TPM_data<-TPM_data %>%relocate(feature)
    if( sum(duplicated(TPM_data$feature)) > 0){
      message("feature (e.g. gene names) should be in the first column. Please remove duplicated feature IDs, Exit!")
      quit(save = "no", status = 1, runLast = FALSE)
    }else{
      rownames(TPM_data) <- TPM_data$feature
      TPM_data = TPM_data[,-1]
      loaded_sample_count <- ncol(TPM_data)
    }
    ####################################################
    #### NA check
    if(sum(is.na(TPM_data)) > 0 ){      
      message(paste0("NA is not allowed in the data, there are ",sum(is.na(TPM_data))," NAs, Exit!"))
      quit(save = "no", status = 1, runLast = FALSE)}
    # #### make sure every expr column is in numeric type
    # matrix_check <- map(TPM_data, is.numeric) %>% unlist
    # if(sum(!matrix_check) > 0){
    #   message("The following column(s) in expression matrix is/are NOT in numeric type. Plese check, Proceed by excluding those samples")
    #   message(paste(names(matrix_check)[!matrix_check], collapse = "; "))
    #   TPM_data = TPM_data[,matrix_check]
    # }
    # message("Gene expression profiles loaded successfully!")
    # message(paste(nrow(TPM_data), "genes and", ncol(TPM_data), "samples are loaded from", $[_input:r], sep = " "))
    
    #### Filter out low exp genes
    keep_genes_idx <- (rowMeans(TPM_data>low_expr_TPM)>low_expr_TPM_percent) 
    TPM_data = TPM_data[keep_genes_idx,]
```

```{r}
log_offset <- 0.0001
TPM_t<-as.data.frame(t(TPM_data))
# df_pca_minus_log<-log10(TPM_t+log_offset)
df_pca_plus_log<-log10(TPM_t+log_offset)

plus.pca<-pca(df_pca_plus_log, center = TRUE, scale =FALSE)
# plus.pca.scaled<-pca(df_pca_plus_log[,1:42339], center = TRUE, scale =TRUE)
# minus.pca<-pca(df_pca_minus_log, center = TRUE, scale =FALSE)
# minus.pca.scaled<-pca(df_pca_minus_log[,1:44883], center = TRUE, scale =TRUE)

plus.pca.plot<-
  Scatter_Density(object = plus.pca,
                  batch = df_pca_plus$batch,
                  title = 'Poly-A plus strand data PCA plot',
                  batch.legend.title = 'batch')
minus.pca.plot<- 
  Scatter_Density(object = minus.pca, 
                  batch = df_pca_minus$batch, 
                  title = 'Poly-A minus strand data PCA plot', 
                  batch.legend.title = 'batch')

```
#Outliers identification method 1: 
Outlier detection by the Mahalanobis distances of PC1 and PC2. Also plot PC1 and PC2 with its confidence ellipse. (package mt)

```{r}
#install.packages('mt')
library('mt')
```

```{r}
#Minus PCA outliers detection
x<-pca.outlier(df_pca_minus_log, center = TRUE, scale =FALSE, conf.level = 0.975) 
x$outlier
x$plot
#Comment: Looks nice, actually got 2 more outliers than the 2 sd away and finding overlaps. Could be a handy way to identify outliers
```

```{r}
p<-pca.outlier(df_pca_plus_log, center = TRUE, scale =FALSE, conf.level = 0.975) 
p$outlier
p$plot
```

#Prepare for RLE AND d-STATISTICS

```{r}
library(RColorBrewer)
    library(ape)
    library(reshape2)
    library(dplyr)
    library(readr)
    ## Setting parameters
    RLEFilterPercent <-0.05
    DSFilterPercent <- 0.05
    ### Hcluster parameter
    pvalues.cut <- 0.05
    topk_genes <- 100
    cluster_percent <- 0.6
    treesNum <- 5
    ## Define functions to accomodate rank deficient covariance matrices in https://github.com/cumc/xqtl-pipeline/issues/307
    mahalanobis = function (x, center, cov, inverted = FALSE, ...) 
        {
            x <- if (is.vector(x)) 
                matrix(x, ncol = length(x))
            else as.matrix(x)
            if (!isFALSE(center)) 
                x <- sweep(x, 2L, center)
            if (!inverted) 
                cov <- MASS::ginv(cov, ...)
            setNames(rowSums(x %*% cov * x), rownames(x))
        }
          
```

#RLE
```{r}
    ## load Data
    # TPM_data <- read_tsv('/restricted/projectnb/casa/skandoi/bulk_qc/fastq.rnaseqc.low_expression_filtered.tpm.gct.gz', col_names = TRUE, comment = "#")%>%as.data.frame()
    # rownames(TPM_data) <- TPM_data$gene_ID
    # TPM_data = TPM_data[,-1]
    RLEFilterLength <- RLEFilterPercent*ncol(TPM_data)
    DSFilter <- DSFilterPercent*ncol(TPM_data)
## Outlier detection
    ### RLE
    # https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L71
    log_offset <- 0.0001
    logtpm = log10(TPM_data%>%as.matrix + log_offset)
    rle=logtpm-apply(logtpm, 1, median) # change "/" to "-" so that we got log(fold-change) which centered on 0 on the RLE plot.
    iqr = apply(rle,2,IQR)
    rle=melt( rle, variable.name = "Sample",value.name ="TPM", id="ID")
    names(rle) <- c("feature","Sample","TPM")
    rle_IQR <- rle %>% group_by(Sample) %>% summarise(IQR = IQR(TPM))
    rle_IQR_range <- rle_IQR$IQR %>% range %>% abs() %>% max()
    rle_IQR_range <- 2*rle_IQR_range %>% ceiling()
    bymedian <- with(rle, reorder(Sample, TPM, IQR))  # sort by IQR
    par(mar=c(3,3,3,3))
    pdf(file = "$[_output:n].RLEplot.pdf")
    boxplot(TPM ~ bymedian, data=rle, outline=F, ylim = c(-rle_IQR_range, rle_IQR_range), las=2, boxwex=1, col='gray', cex.axis=0.3, main="RLE plot before QC", xlab="", ylab="Residual expression levels", frame=F)
    dev.off()
    ExpPerSample <- nrow(TPM_data)
    #get the outlier list using 
    RLEfilter_start<-as.integer(ncol(TPM_data)-round(RLEFilterLength)+1)
    RlEfilter_end<-as.integer(ncol(TPM_data))
    RLEFilterList <- levels(bymedian)[RLEfilter_start:RlEfilter_end]
    RLEFilterList <- as.character(RLEFilterList)
    message(paste0("The right most ", RLEFilterPercent*100, "% samples (N = ", length(RLEFilterList), ") are marked as candidate outliers in this step:") )
    message(RLEFilterList)
```


```{r}
### D-s
    sampleDists <- 1 - cor(logtpm, method='spearman')
    hc <- hclust(as.dist(sampleDists), method = "complete")
    hcphy <- as.phylo(hc)
    sampleDists <- 1 - cor(logtpm, method='spearman')
    hc <- hclust(as.dist(sampleDists), method = "complete")
    hcphy <- as.phylo(hc)
    D = apply(1-sampleDists, 1, median)
    pdf(file = "$[_output:n].D_stat_hist.pdf")
    hist(D, breaks=100, ylab="Number of samples", xlab="D-statistic", main="Histogram of Sample D-statistics before data QC")
    dev.off()
    
    DSFilter <- sort(D)[DSFilter]
    D<-as.data.frame(D)
    D<-data.frame(Sample = rownames(D),D = D$D)
    D_filterList = D%>%filter(D <= DSFilter)
    D_filterList <- D_filterList$Sample
    D_filterList<-as.character(D_filterList)
    message(paste0("The right most ", DSFilterPercent*100, "% samples (N=", length(D_filterList), ") are marked as candidate outliers in this step:") )
    message(D_filterList)
```
